source/module_hsolver/diago_elpa.cpp

```diff
@@ -120,35 +122,49 @@ void DiagoElpa<double>::diag(hamilt::Hamilt<double>* phm_in, psi::Psi<double>& p
     std::vector<double> eigen(GlobalV::NLOCAL, 0.0);
 
     bool isReal = true;
-    MPI_Comm COMM_DIAG = setmpicomm(); // set mpi needed;
-    // ELPA_Solver es(isReal, COMM_DIAG, GlobalV::NBANDS, h_mat.row, h_mat.col, h_mat.desc);
-    ELPA_Solver es((const bool)isReal, COMM_DIAG, (const int)GlobalV::NBANDS, (const int)h_mat.row,
-                   (const int)h_mat.col, (const int*)h_mat.desc);
+    MPI_Comm COMM_DIAG = MPI_COMM_NULL;
+    if (Parallel_K2D<double>::get_instance().get_initialized()) {
+        COMM_DIAG = Parallel_K2D<double>::get_instance().POOL_WORLD_K2D;
```

Comment:
It is not good to use Global class here, the communicator should be passed into HSolver.

Reply:
I absolutely agree with this comment and think it is ugly to use Global class here. In the meantime, removing this usage of global class means we need to refactor the interface `virtual void diag(hamilt::Hamilt<T, Device>* phm_in, psi::Psi<T, Device>& psi, Real* eigenvalue_in)` of the base class `diagH`, which requires significant new refactorization of the whole `hsolver` module ...